rs/index_server/src/metrics.rs

```diff
@@ -0,0 +1,118 @@
+use std::io;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use futures::future::BoxFuture;
+use http_body_util::{combinators, BodyExt, Full};
+use hyper::body::{Bytes, Incoming};
+use hyper::server::conn::http1;
+use hyper::service::service_fn;
+use hyper::{Request, Response};
+use hyper_util::rt::TokioIo;
+use prometheus_client::encoding::text::encode;
+use prometheus_client::metrics::counter::Counter;
+use prometheus_client::registry::Registry;
+use tokio::net::TcpListener;
+use tokio::pin;
+use tokio::signal::unix::{signal, SignalKind};
+
+pub struct MetricsServer {
+    registry: Registry,
+    request_counter: Counter<u64>,
+}
+
+impl MetricsServer {
+    /// Create a new instance of the metrics server, with a request counter metric.
+    pub fn new() -> Self {
+        let request_counter: Counter<u64> = Default::default();
+
+        let mut registry = <Registry>::with_prefix("index_server_metrics");
+
+        registry.register(
+            "requests",
+            "How many requests to the metrics endpoint have been made",
+            request_counter.clone(),
+        );
+        Self {
+            registry,
+            request_counter,
+        }
+    }
+
+    /// Start a HTTP server to report metrics.
+    pub async fn serve(self, addr: SocketAddr) -> io::Result<()> {
+        let registry = Arc::new(self.registry);
```

Comment:
don't transfer the ownership of `registry` from `self` to `Arc`. This is very anti-patern. Also, why are you passing `self` without `ref`? Are you sure?